server/api/reviews.go

```diff
@@ -17,8 +23,180 @@ type ReviewBody struct {
+
+	categoriesFilter := bson.M{"categories": bson.M{"$in": categoriesArr}}
+
+	if len(searchQuery) > 0 && len(categoriesQuery) > 0 {
+		reviewsCursor, err = DB.Collection(db.ReviewsCollection).Find(db.Ctx, categoriesFilter, opts)
+		return search(reviewsCursor, err, searchQuery)
+	} else if len(searchQuery) > 0 {
+		reviewsCursor, err = DB.Collection(db.ReviewsCollection).Find(db.Ctx, bson.M{}, opts)
+		return search(reviewsCursor, err, searchQuery)
+	} else if len(categoriesQuery) > 0 {
+		reviewsCursor, err = DB.Collection(db.ReviewsCollection).Find(db.Ctx, categoriesFilter, opts)
+	} else {
+		reviewsCursor, err = DB.Collection(db.ReviewsCollection).Find(db.Ctx, bson.M{}, opts)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err = reviewsCursor.All(db.Ctx, &reviews); err != nil {
+		return nil, err
+	}
+
+	return reviews, nil
+}
+
+func search(cursor *mongo.Cursor, err error, search string) ([]db.Review, error) {
+
+	var reviews []db.Review
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err = cursor.All(db.Ctx, &reviews); err != nil {
+		return nil, err
+	}
+
+	searchKeywords := strings.Split(search, " ")
+
+	reviewsWithKeyword := make(map[primitive.ObjectID]db.Review)
+
+	// Review body should contain matched keyword atleast once
+	for _, item := range searchKeywords {
+		for _, entry := range reviews {
+			if contains(strings.Split(strings.ToLower(strings.Trim(entry.Text, " ")), " "), item) {
+				reviewsWithKeyword[entry.ID] = entry
+			}
+		}
+	}
+
+	withKeywordsReviews := make([]db.Review, 0)
+
+	for _, value := range reviewsWithKeyword {
+		withKeywordsReviews = append(withKeywordsReviews, value)
+	}
+
+	occurences := make(map[primitive.ObjectID]map[string]int)
+
+	//Find occurence of keyword in each review
+	for _, item := range withKeywordsReviews {
+		for _, entry := range searchKeywords {
+			reviewBodyArray := strings.Split(strings.ToLower(strings.Trim(item.Text, " ")), " ")
+			for _, reviewItem := range reviewBodyArray {
+				if contains(reviewBodyArray, entry) {
+					_, ok := occurences[item.ID]
+
+					if !ok {
+						occurences[item.ID] = make(map[string]int)
+					}
+					if reviewItem == entry {
+						occurences[item.ID][reviewItem] += 1
+					}
+				}
+			}
+		}
+	}
+
+	if len(occurences) <= 0 {
+		return reviews, nil
+	}
+
+	occurenceMap := make(map[primitive.ObjectID]int)
+
+	for index, item := range occurences {
+		for _, entry := range item {
+			occurenceMap[index] = entry
+		}
+	}
+
+	sortReviews(&withKeywordsReviews, occurenceMap)
+
+	return withKeywordsReviews, nil
+}
+
+type SortableReview struct {
+	reviews   []db.Review
+	frequency map[primitive.ObjectID]int
+}
+
+func (s SortableReview) Len() int {
+	return len(s.reviews)
+}
+
+func (s SortableReview) Less(i, j int) bool {
+	less := s.frequency[s.reviews[i].ID] > s.frequency[s.reviews[j].ID]
+	if s.frequency[s.reviews[i].ID] == s.frequency[s.reviews[j].ID] {
+		less = s.reviews[i].Likes > s.reviews[j].Likes
```

Comment:
Here, only sort by `likes` is considered, but sort by recency and book name are not.

Reply:
When frequencies is the same for two reviews, it should be sort by `likes`, `created_at` or `book_name` depends on the query parameters, here only `likes` is considered.